src/pip/_internal/models/format_control.py

```diff
@@ -23,7 +27,14 @@ def __init__(self, no_binary=None, only_binary=None):
 
     def __eq__(self, other):
         # type: (object) -> bool
-        return self.__dict__ == other.__dict__
+        if isinstance(other, self.__class__):
+            if self.__slots__ == other.__slots__:
+                attr_getters = [operator.attrgetter(attr)
+                                for attr in self.__slots__]
+                return all(getter(self) == getter(other)
+                           for getter in attr_getters)
+
+        return False
```

Comment:
nit-ish: I like short-circuiting to avoid additional indentation. Additionally, we'd also want to let other classes implement equality if they want to, so returning NotImplemented allows for a tiny bit more flexibility (which we might not use to be fair).

```suggestion
        if not isinstance(other, self.__class__):
            return NotImplemented
            
        if self.__slots__ != other.__slots__:
            return NotImplemented

        return all(
            getattr(self, k) == getattr(other, k)
            for k in self.__slots__
        )
```